text/0000-router-service.md

```diff
@@ -0,0 +1,260 @@
+- Start Date: 2015-09-24
+- RFC PR:
+- Ember Issue:
+
+# Summary
+
+This RFC proposes:
+
+ - creating a public `router` service that is a superset of today's `Ember.Router`.
+ 
+ - codifying and expanding the supported public API for the `transition` object that is currently passed to `Route` hooks.
+
+These topics are closely related because they share a unified `RouteInfo` type, which will be described in detail.
+
+# Motivation
+
+Given the modern Ember concepts of Components and Services, it is clear that routing capability should be exposed as a Service. I hope this is uncontroversial, given that we already implement it as a service internally, and given that usage of these nominally-private APIs is already becoming widespread.
+
+The immediate benefit of having a `RouterService` is that you can inject it into components, giving them a friendly way to initiate transitions and ask questions about the current global router state.
+
+A second benefit is that we have the opportunity to add new capabilities to the `RouterService` to replace several common patterns in the wild that dive into private internals in order to get things done. There are several places where we leak internals from router.js, and we can plug those leaks.
+
+# Detailed design
+
+## RouterService
+
+By way of a simple example, the router service behaves like this:
+
+```js
+import Component from 'ember-component';
+import service from 'ember-service/inject';
+
+export default Component.extend({
+  router: service(),
+  actions: {
+    goToMars() {
+      this.get('router').transitionTo('planet.mars');
+    }
+  }
+});
+```
+
+Like any Service, it can also be injected into Helpers, Routes, etc.
+
+### Relationship between EmberRouter and RouterService
+
+Q: "Why are you calling this thing 'router' when we already have a router? Shouldn't the new thing be called 'routing' or something else?".
+
+A: We shouldn't have two things. From the user's perspective, there is just "the router", and it happens to be available as a service. While we're free to continue implementing it as multiple classes under the hood, the public API should present as a single, coherent concept.
+
+Terminology:
+
+ - `EmberRouter` is the class that we already have today, defined in `ember-routing/system/router` and available publicly as `Ember.Router`
+ - `RouterService` is the new class I am proposing.
+
+`EmberRouter` has the following public API today:
+
+ - `map`
+ - `location`
+ - `rootURL`
+ - `willTransition`
+ - `didTransition`
```

Comment:
No, one of the explicit goals of this RFC is to codify public API so that people can stop messing around with handlerInfos (which are an internal implementation detail of router.js and are not Ember public API).